phoenix-core/src/main/java/org/apache/phoenix/expression/LiteralExpression.java

```diff
@@ -57,181 +56,236 @@
-                } else {
-                    b = StringUtil.padChar(b, maxLength);
-                }
-            } else if (value != null) {
-                maxLength = ((String)value).length();
+
+        private void setDefaults() {
+            if (this.determinism == null) {
+                this.determinism = Determinism.ALWAYS;
+            }
+            if (this.sortOrder == null) {
+                this.sortOrder = SortOrder.getDefault();
+            }
+            if (this.rowKeyOrderOptimizable == null) {
+                this.rowKeyOrderOptimizable = true;
             }
-        } else if (type.isArrayType()) {
-            maxLength = ((PhoenixArray)value).getMaxLength();
-        }
-        if (b.length == 0) {
-            return getTypedNullLiteralExpression(type, determinism);
-        }
-        if (maxLength == null) {
-            maxLength = type.isFixedWidth() ? type.getMaxLength(value) : null;
         }
-        return new LiteralExpression(value, type, b, maxLength, scale, sortOrder, determinism);
-    }
 
-    public LiteralExpression() {
-    }
-    
-    public LiteralExpression(byte[] byteValue) {
-        this.byteValue = byteValue!=null ? byteValue : ByteUtil.EMPTY_BYTE_ARRAY;
-        this.determinism = Determinism.ALWAYS;
-    }
+        /**
+         * Build method for Literal Expressions that have only fields value and determinism (can be null)
+         * or only field byteValue
+         * @param byteValueOnly true if LiteralExpression's only field is byteValue; false otherwise
+         * @return LiteralExpression object
+         */
+        public LiteralExpression buildValueAndDeterminism(boolean byteValueOnly) {
+            if (byteValueOnly) {
+                return new LiteralExpression(null, null, this.byteValue = this.byteValue != null
+                        ? this.byteValue : ByteUtil.EMPTY_BYTE_ARRAY, null, null, SortOrder.ASC,
+                        Determinism.ALWAYS);
+            }
+            setDefaults();
+            if (this.value instanceof Boolean) {
+                return LiteralExpression.getBooleanLiteralExpression((Boolean)this.value,
+                        this.determinism);
+            } else if (this.value == null) {
+                return LiteralExpression.getNullLiteralExpression(this.determinism);
+            }
+            PDataType type = PDataType.fromLiteral(this.value);
+            this.byteValue = type.toBytes(this.value);
+            if (type.isNull(this.byteValue)) {
+                return LiteralExpression.getTypedNullLiteralExpression(type, this.determinism);
+            }
+            if (type == PVarchar.INSTANCE) {
+                String s = (String) this.value;
+                if (s.length() == this.byteValue.length) { // single byte characters only
+                    type = PChar.INSTANCE;
+                }
+            }
+            return new LiteralExpression(this.value, type, this.byteValue, type == null
```

Comment:
Same here. Why do we need to call a constructor of this any longer? We ideally shouldn't have any more such constructors. We should just have 1 constructor that has a `Builder` object as a parameter (and which is private and only called when `build()` is called). Take a look at how this is done in PTableImpl.